Create rule S7205: Dependencies should be verified against a known checksum or signature (#4702)
SonarSource/rspec: rules/S7205/kotlin/rule.adoc

> Software projects often rely on external code libraries, known as dependencies. Package managers such as Gradle, allow developpers to reference dependencies for their projects.

Reviewer: ```suggestion
Software projects often rely on external code libraries, known as dependencies. Package managers, such as Gradle, allow developers to reference dependencies for their projects.
```

These dependencies simplify development, but also introduce risk as they download and include external code based on a project's configuration. 
Integrity checking is the step of verifying that the downloaded or included dependency code is exactly what the developer expects. Without this verification, the application cannot guarantee that the dependency is legitimate.

== Why is this an issue?

> Failing to verify the integrity of dependencies before using them is a significant security problem. It exposes your application, and potentially your users, to several risks. The core issue is that you're running code from an untrusted source without checking it, effectively giving an attacker a direct pathway into your application.

Reviewer: ```suggestion
Failing to verify the integrity of dependencies before using them is a significant security problem. It exposes your application, and potentially your users, to several risks. The core issue is that you are running code from an untrusted source without checking it, effectively giving an attacker a direct pathway into your application.
```

As per styling guide for RSPEC.

> This is often a key component of what's called a "supply chain attack." The attacker isn't directly attacking your application. Instead, they're attacking a component you use. This is an important consideration because the attack's source is less obvious. You might diligently secure your own code, but overlook the risk introduced by external dependencies.

Reviewer: ```suggestion
This is often a key component of what's called a "supply chain attack." The attacker is not directly attacking your application. Instead, they're attacking a component you use. This is an important consideration because the attack's source is less obvious. You might diligently secure your own code, but overlook the risk introduced by external dependencies.
```

As per styling guide for RSPEC.

== How to fix it in Gradle

=== Code examples

==== Noncompliant code example

[source,kotlin,diff-id=1,diff-type=noncompliant]
----
dependencies {
    implementation("com.example:a-dependency:1.0")
}

configurations { 
    all {
        resolutionStrategy {
            disableDependencyVerification() // Noncompliant: dependency verification is disabled
        }
    }
}
----

Absence of a `verification-metadata.xml` file in the `gradle` directory of your project will also result in the Gradle build not verifying the integrity of the dependencies.

==== Compliant solution

[source,kotlin,diff-id=1,diff-type=compliant]
----
dependencies {
    implementation("com.example:adependency:1.0")
}
----

Create a `verification-metadata.xml` in the `gradle` directory of your project.

> Use `./gradlew --write-verification-metadata pgp,sha256 --export-keys` to bootstrap the file content

Reviewer: ```suggestion
Use `./gradlew --write-verification-metadata pgp,sha256 --export-keys` to bootstrap the file content.
```
